### PR TITLE
Commands json schema

### PR DESCRIPTION
**Pull request by hansl**

### CODE REVIEW

**Comment by hansl** [comment]
This is a follow up and replaces #11742

**Comment by hansl** [comment]
Will ignore CLA Bot as Mike signed it at the time of his PR, and I obviously signed it too :)

**Comment by petebacondarwin** [comment]
Why are none of the schema properties "required"?

**Comment by hansl** [comment]
Smart defaults (the `$default` property) don't work well right now with required :/

File: packages/angular/cli/commands/generate.ts [hunk 1]
```diff
@@ -60,7 +48,8 @@ export class GenerateCommand extends SchematicCommand {
     const [collectionName, schematicName] = this.parseSchematicInfo(options);
 
     // remove the schematic name from the options
-    options._ = options._.slice(1);
+    // options._ = options._.slice(1);
```
Comment (filipesilva) [inline packages/angular/cli/commands/generate.ts]: Leftover commented code.

File: packages/angular/cli/models/command.ts [hunk 2]
```diff
@@ -51,47 +55,76 @@ export abstract class Command<T = any> {
...
+      this.logger.info(description);
+    }
+    this.printHelpUsage(commandName, this.options);
     this.printHelpOptions(this.options);
   }
 
-  protected printHelpUsage(name: string, args: string[], options: Option[]) {
+  private getArgIndex(def: OptionSmartDefault | undefined): number {
+    if (def === undefined || def.index === undefined || typeof def.index !== 'number') {
+      return 99999;
```
Comment (filipesilva) [inline packages/angular/cli/models/command.ts]: I don't think we'll get to 99999 args but this looks like it should be a non-number type, probably `null`.

File: packages/angular_devkit/schematics/tools/index.ts [hunk 3]
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 export * from './description';
+export * from './export-ref';
```
Comment (filipesilva) [inline packages/angular_devkit/schematics/tools/index.ts]: This change to the public API should be a feat in it's own commit.

File: packages/angular/cli/models/command-runner.ts [hunk 4]
```diff
@@ -70,18 +101,27 @@ export async function runCommand(commandMap: CommandMap,
     ? CommandScope.inProject
     : CommandScope.outsideProject;
 
-  let Cmd: CommandConstructor | null;
-  Cmd = commandName ? findCommand(commandMap, commandName) : null;
+  const commandMapPath = findUp('commands.json', __dirname);
```
Comment (qiyigg) [inline packages/angular/cli/models/command-runner.ts]: Possible to make the config file path as a parameter?

File: packages/angular_devkit/schematics/tools/index.ts (hunk 3 above)
Comment (hansl) [reply packages/angular_devkit/schematics/tools/index.ts]: `ExportRef` should be part of core anyway.

File: packages/angular/cli/commands/generate.ts (hunk 1 above)
Comment (hansl) [reply packages/angular/cli/commands/generate.ts]: Done.

File: packages/angular_devkit/schematics/tools/index.ts (hunk 3 above)
Comment (hansl) [reply packages/angular_devkit/schematics/tools/index.ts]: It's in a feature. This was from Mike's PR so I don't want to refactor it. I'll refactor it later (would like to move it out of schematics).

File: packages/angular/cli/models/command-runner.ts (hunk 4 above)
Comment (hansl) [reply packages/angular/cli/models/command-runner.ts]: Done.

File: packages/angular/cli/commands/add.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+  "$schema": "http://json-schema.org/schema",
+  "id": "AddCommandOptions",
+  "description": "Add support for a library to your project.",
+  "$longDescription": "",
+
+  "$scope": "in",
+  "$impl": "./add-impl#AddCommand",
```
Comment (qiyigg) [inline packages/angular/cli/commands/add.json]: It seems we still have to maintain a bunch of ${command}.json files internally. I don't know how often will these json files change? Will it make sense for us to maintain a different set of json files?

Comment (hansl) [reply packages/angular/cli/commands/add.json]: I want to unify schematics, architect and cli json schema files. There is already one code path but I want to make it more streamlined. This could also open up the way to custom commands down the road when we want to.

**Comment by hansl** [comment]
@googlebot Please rerun.
